tests/lib/sample_rate_converter/src/main.c

```diff
@@ -0,0 +1,1065 @@
+	zassert_equal(conv_ctx.output_sample_rate, output_sample_rate,
+		      "Output sample rate not as expected");
+	zassert_equal(conv_ctx.conversion_ratio, conversion_ratio,
+		      "Conversion ratio not as expected");
+	zassert_equal(conv_ctx.conversion_direction, CONVERSION_DIR_DOWN,
+		      "Conversion direction is not as expected");
+	zassert_equal(conv_ctx.filter_type, filter, "Filter set incorrectly");
+	zassert_equal(conv_ctx.input_buf.bytes_in_buf, 0, "Bytes in input buffer not as expected");
+	zassert_equal(ring_buf_size_get(&conv_ctx.output_ringbuf), 0,
+		      "Number of bytes in output ringbuffer not as expected %d",
+		      ring_buf_size_get(&conv_ctx.output_ringbuf));
+
+	zassert_equal(output_written, expected_output_samples * sizeof(uint16_t),
+		      "Output size was not as expected (%d)", output_written);
+
+	/* Verify output bytes, expected three input samples per output sample */
+	for (int i = 0; i < expected_output_samples; i++) {
+		zassert_within(output_samples[i], input_samples[i * 3], 1100);
+	}
+}
+
+ZTEST(suite_sample_rate_converter, test_init_valid_interpolate_24khz_16bit)
+{
+	int ret;
+
+	uint32_t input_sample_rate = 24000;
+	uint32_t output_sample_rate = 48000;
+	uint32_t conversion_ratio = output_sample_rate / input_sample_rate;
+	enum sample_rate_converter_filter filter = SAMPLE_RATE_FILTER_SIMPLE;
+
+	uint16_t input_samples[] = {2000, 4000, 6000, 8000, 10000, 12000};
+	size_t num_samples = ARRAY_SIZE(input_samples);
+	size_t expected_output_samples = num_samples * conversion_ratio;
+	uint16_t output_samples[expected_output_samples];
+	size_t output_written;
+
+	ret = sample_rate_converter_process(
+		&conv_ctx, filter, input_samples, num_samples * sizeof(uint16_t), input_sample_rate,
+		output_samples, expected_output_samples * sizeof(uint16_t), &output_written,
+		output_sample_rate);
+
+	zassert_equal(ret, 0, "Sample rate conversion process failed");
+	zassert_equal(conv_ctx.input_sample_rate, input_sample_rate,
+		      "Input sample rate not as expected");
+	zassert_equal(conv_ctx.output_sample_rate, output_sample_rate,
+		      "Output sample rate not as expected");
+	zassert_equal(conv_ctx.conversion_ratio, conversion_ratio,
+		      "Conversion ratio not as expected");
+	zassert_equal(conv_ctx.conversion_direction, CONVERSION_DIR_UP,
+		      "Conversion direction is not as expected");
+	zassert_equal(conv_ctx.filter_type, filter, "Filter set incorrectly");
+	zassert_equal(conv_ctx.input_buf.bytes_in_buf, 0, "Bytes in input buffer not as expected");
+	zassert_equal(ring_buf_size_get(&conv_ctx.output_ringbuf), 0,
+		      "Number of bytes in output ringbuffer not as expected %d",
+		      ring_buf_size_get(&conv_ctx.output_ringbuf));
+
+	zassert_equal(output_written, expected_output_samples * sizeof(uint16_t),
+		      "Output size was not as expected (%d)", output_written);
+
+	/* Verify output bytes, expect two samples in output per input */
+	for (int i = 0; i < expected_output_samples; i++) {
```

Comment:
It proved difficult to find a good way to check the actual output values, as a small number of samples really isn't enough for the filter to produce good output. So for this test case I found that a sanity check could be that the output samples aren't higher than the input samples (since the filter is a low pass filter). We can of course hard code the expected output values, but IMO that doesn't give us any more confidence and we get tests that are harder to maintain.

Reply:
OK, let's discuss